### PR TITLE
Multi-Scale Slice Attention: coarse+fine parallel resolution

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Surface pressure has **multi-scale structure**: low-frequency components capture the global Cp distribution (lift, drag, pressure recovery) while high-frequency components capture the leading-edge suction peak and trailing-edge separation. The Transolver uses a single slice resolution (96 slices) — but 96 slices may be too coarse to resolve the LE suction spike, while simultaneously too fine-grained for efficient global inter-foil coupling.

**Multi-scale slice attention** runs two parallel Transolver block stacks at different resolutions:
- **Coarse branch (48 slices):** Each slice covers more physical area → better global context and inter-foil coupling (fore-aft wake interaction). Fewer slices = faster per-layer computation.
- **Fine branch (192 slices):** Each slice covers a smaller area → better local resolution for LE suction peak, TE separation, boundary layer features.

The two branches share the input projection but have separate TransolverBlock stacks. Their outputs are fused (concatenated + linear projection) before the SRF/output head. This is the **Feature Pyramid Network (FPN)** / **U-Net** concept applied to slice attention — a proven multi-scale strategy in computer vision (He et al. 2017) now applied to mesh-based neural operators.

**Why this should work:** The current 96-slice resolution is a compromise. At 96 slices with ~27k nodes, each slice handles ~280 nodes — not fine enough for the ~50 nodes around the LE suction spike, not coarse enough for the ~800 nodes across both foils in tandem. Multi-scale decouples these two needs.

**Target:** p_tan (coarse branch captures inter-foil coupling), p_in (fine branch captures LE spike).

### 1. Add flags
```python
parser.add_argument('--multi_scale_slice', action='store_true',
    help='Dual-resolution slice attention: coarse (48) + fine (192) branches')
parser.add_argument('--slice_num_coarse', type=int, default=48,
    help='Coarse branch slice count (default: 48)')
parser.add_argument('--slice_num_fine', type=int, default=192,
    help='Fine branch slice count (default: 192)')
```

### 2. Architecture changes

When `--multi_scale_slice` is active, create TWO parallel TransolverBlock stacks:

```python
if args.multi_scale_slice:
    # Coarse branch: fewer slices, same hidden dim, n_layers layers
    self.blocks_coarse = nn.ModuleList([
        TransolverBlock(
            num_heads=num_heads,
            hidden_dim=hidden_dim,
            slice_num=args.slice_num_coarse,
            # ... same params as current blocks
        ) for _ in range(args.n_layers)
    ])
    # Fine branch: more slices, same hidden dim, n_layers layers
    self.blocks_fine = nn.ModuleList([
        TransolverBlock(
            num_heads=num_heads,
            hidden_dim=hidden_dim,
            slice_num=args.slice_num_fine,
            # ... same params as current blocks
        ) for _ in range(args.n_layers)
    ])
    # Fusion: concat coarse + fine, project back to hidden_dim
    self.scale_fusion = nn.Sequential(
        nn.Linear(hidden_dim * 2, hidden_dim),
        nn.GELU(),
        nn.Linear(hidden_dim, hidden_dim),
    )
else:
    # Standard single-resolution blocks (current behavior)
    self.blocks = nn.ModuleList([...])
```

### 3. Forward pass

```python
if self.args.multi_scale_slice:
    h_input = self.preprocess(x)  # shared input projection
    
    # Coarse branch: 48 slices
    h_coarse = h_input.clone()
    for block in self.blocks_coarse:
        h_coarse = block(h_coarse, pos)  # pos = spatial coordinates for slice routing
    
    # Fine branch: 192 slices
    h_fine = h_input.clone()
    for block in self.blocks_fine:
        h_fine = block(h_fine, pos)
    
    # Fuse: concatenate and project
    h = self.scale_fusion(torch.cat([h_coarse, h_fine], dim=-1))  # [B, N, D]
else:
    # Standard forward (unchanged)
    h = self.preprocess(x)
    for block in self.blocks:
        h = block(h, pos)
```

### 4. Memory and compute considerations

- 2× TransolverBlock stacks doubles the backbone parameter count (~2M total)
- Coarse (48 slices) and fine (192 slices) may have different memory profiles
- The branches are **independent** and could run in parallel (but sequential is fine for correctness)
- Monitor VRAM usage — if >90GB, try reducing `--n_layers` to 2 for BOTH branches
- If 2× n_layers=3 is too slow (fewer epochs in 30 min), try n_layers=2 for each branch (still 4 total TransolverBlocks, same as current 3+1 SRF)

### 5. Alternative: Shared early layers, split late layers

If full dual-stack is too expensive, try a hybrid:
```python
# Shared first 2 layers at 96 slices
h = block_shared_1(h, pos)
h = block_shared_2(h, pos)

# Split: coarse (1 layer, 48 slices) + fine (1 layer, 192 slices)
h_coarse = block_coarse(h, pos)
h_fine = block_fine(h, pos)
h = self.scale_fusion(torch.cat([h_coarse, h_fine], dim=-1))
```

This reduces parameter overhead to just 1 extra TransolverBlock.

### Run commands

```bash
# Seed 42
CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent tanjiro --wandb_name "tanjiro/multi-scale-s42" --wandb_group multi-scale-slice \
  --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature \
  --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only \
  --cp_panel_scale 0.1 --wake_angle_feature --vortex_panel_velocity \
  --vortex_panel_scale 0.1 --vortex_panel_n 64 \
  --multi_scale_slice --slice_num_coarse 48 --slice_num_fine 192

# Seed 73 (same flags, --seed 73, --wandb_name "tanjiro/multi-scale-s73")
```

**Important:** If full dual-stack exceeds 90GB VRAM or causes timeout, fall back to the shared-early-split-late variant with `--n_layers 2` per branch. Report VRAM and epoch timing in your results.

## Baseline

| Metric | 2-seed avg | Target to beat |
|--------|-----------|----------------|
| **p_in** | **11.872** | < 11.872 |
| p_oodc | 7.459 | < 7.459 |
| **p_tan** | **26.319** | < 26.319 |
| **p_re** | **6.229** | < 6.229 |

W&B baseline: aycq1m8m (seed 42), 9sk276v6 (seed 73)

Reproduce:
```bash
cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output \
  --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature \
  --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT tanjiro:

**Bug fix note (separate from experiment):** Found that `SENPAI_MAX_EPOCHS` and `SENPAI_TIMEOUT_MINUTES` env vars were set in the environment but not read by `train.py` (code used hardcoded `MAX_EPOCHS=500`, `MAX_TIMEOUT=180`). Fixed by reading from env vars with the hardcoded values as defaults.

Starting multi-scale slice attention experiments now. Debug run confirmed 59.7GB VRAM usage (within 97GB budget). Launching seed 42 and seed 73 in parallel on GPUs 1 and 2.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2373 because ADVISOR: Closing based on W&B results from the completed training runs (ktfhbueu seed 73, c8m7l12s seed 42).

## Results (from W&B)

| Metric | s42 | s73 | 2-seed avg | Baseline | Delta |
|--------|-----|-----|-----------|----------|-------|
| p_in | 12.181 | 12.048 | 12.114 | 11.872 | +2.0% |
| p_oodc | 7.650 | 7.591 | 7.620 | 7.459 | +2.2% |
| p_tan | 28.130 | 27.401 | 27.766 | 26.319 | +5.5% |
| p_re | 6.871 | 6.598 | 6.735 | 6.229 | +8.1% |

All metrics above baseline. The multi-scale slice attention approach failed, primarily due to computational overhead: epoch time of 103s (vs 68s baseline) meant only ~105 epochs completed in the 180-min timeout vs 155 for baseline. The model couldn't enter the final cosine fine-tuning phase (T_max=150).

The p_in closeness (+2.0%) shows the multi-scale mechanism doesn't fundamentally hurt the model, but the 50% per-epoch overhead is incompatible with our training budget. p_tan +5.5% and p_re +8.1% confirm no benefit.

Note: tanjiro's pod was restarted and is re-running this experiment — please stop training and wait for new assignment.

Bug fix note: thank you for discovering and fixing the SENPAI_MAX_EPOCHS / SENPAI_TIMEOUT_MINUTES env var issue (alphonse independently found the same bug). This fix should be cherry-picked to the advisor branch..

**Comment by tcapelle** [comment]
STUDENT tanjiro:

## Results

**Setup:**
- Architecture: Multi-scale slice attention with n_layers=2 per branch (2 coarse @ 48 slices + 2 fine @ 192 slices + 1 output head = 5 total TransolverBlocks)
- Why n_layers=2 not 3: n_layers=3 per branch took 250s/epoch → only 7 epochs in 30 min. Fell back to n_layers=2 per PR instructions ("try n_layers=2 for each branch"). n_layers=2 takes ~104s/epoch → 16-17 epochs.
- VRAM: 61.7-61.9 GB (within 97 GB budget)
- Total params: 2,482,323 (main model) + 42,499 (SRF head)

**Commands run:**
```bash
# Seed 42
CUDA_VISIBLE_DEVICES=1 python train.py --agent tanjiro --wandb_name "tanjiro/multi-scale-s42" --wandb_group multi-scale-slice --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 2 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64 --multi_scale_slice --slice_num_coarse 48 --slice_num_fine 192

# Seed 73 (same flags)
CUDA_VISIBLE_DEVICES=2 python train.py --agent tanjiro --wandb_name "tanjiro/multi-scale-s73" --wandb_group multi-scale-slice --seed 73 ...
```

**Surface MAE at best epoch (16-17 epochs total, timeout at 30 min):**

| Metric | Seed 42 | Seed 73 | 2-seed avg | Baseline (ep.155) |
|--------|---------|---------|------------|-------------------|
| p_in | 51.3 | 47.1 | **49.2** | 11.872 |
| p_tan | 61.0 | 66.7 | **63.9** | 26.319 |
| p_oodc | 33.3 | 32.7 | **33.0** | 7.459 |
| p_re | 25.6 | 29.3 | **27.5** | 6.229 |

W&B: 3hdijxuj (seed 42, best ep. 16), hffax88j (seed 73, best ep. 17)
Peak memory: ~62 GB

**⚠️ These results are NOT comparable to baseline.** The baseline ran 155 epochs under 180-min limit. These experiments stopped at 16-17 epochs under the 30-min SENPAI_TIMEOUT_MINUTES limit. The model was still rapidly improving when timeout hit (val_loss: 9 at ep.1 → 1.5 at ep.16, no plateau).

**What happened:**
- Multi-scale model takes 1.7× longer per epoch (104s vs ~70s for baseline)
- n_layers=3 per branch was 2.5× slower (250s/epoch → only 7 epochs), so fell back to n_layers=2
- Even n_layers=2 limits to 16-17 epochs in 30 minutes
- Cannot conclude whether multi-scale helps — epoch 16 vs 155 is not a fair comparison

## Research

Key insight: Transolver slice attention is a **low-rank kernel approximation** (rank=K). Running K=48 and K=192 in parallel gives complementary spectral decompositions — not different spatial receptive fields. K=48 produces smoother, globally-coupled features; K=192 produces sharper representations. DPNO (2024) shows dual-path operators help for PDEs with sharp gradients (on regular grids). Whether this spectral decomposition benefits LE suction spike vs. inter-foil coupling as hypothesized — could not verify in the truncated runs.

## Bug fixes

1. **`SENPAI_MAX_EPOCHS`/`SENPAI_TIMEOUT_MINUTES` not read by train.py**: These env vars were set but train.py hardcoded `MAX_EPOCHS=500` and `MAX_TIMEOUT=180`. Fixed to read from env with fallback to hardcoded defaults. This explains why prior experiments ran 155 epochs (180 min) instead of hitting the 50-epoch/30-min limit.

2. **Verification NaN**: End-of-training verification produces NaN Pipeline/Manual MAE. Appears to be a pre-existing issue: early-training models may produce extreme outputs that overflow `torch.sinh()` (no clamp before sinh in verification paths). Live training metrics are unaffected.

## Suggested follow-ups

1. **Extended time budget**: Multi-scale needs 90-150 min to approach convergence. Cannot evaluate under 30-min limit.
2. **Shared early + split late**: 2 shared blocks (96 slices) + 1 coarse + 1 fine + output head — reduces branch redundancy per research literature.
3. **Gated fusion**: Replace concat+linear with learned gate (prevents one branch from dominating).
4. **Fix verification NaN**: Add `clamp(-20, 20)` before `torch.sinh()` in verification paths A/B/C.
